src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp

```diff
@@ -193,8 +193,8 @@ void MergeTreeWhereOptimizer::analyzeImpl(Conditions & res, const ASTPtr & node,
             /// Condition depend on some column. Constant expressions are not moved.
             !cond.identifiers.empty()
             && !cannotBeMoved(node, is_final)
-            /// Do not take into consideration the conditions consisting only of the first primary key column
-            && !hasPrimaryKeyAtoms(node)
+            /// when use final, do not take into consideration the conditions consisting of column that not belong to the sorting key columns
```

Comment:
I can't understand this: why we will not consider non-sorting columns when using final? Any correctness issues? @KochetovNicolai

Reply:
Yes, this is a correctness issue.
For queries with FINAL we may want to merge some rows with the same sorting key values. So, we cannot filter (except if the condition is fully on top of sorting key) before FINAL happens (=cannot move the condition to prewhere)